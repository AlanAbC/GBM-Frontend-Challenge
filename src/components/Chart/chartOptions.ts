import dayjs from "dayjs";
import { ColorType } from "lightweight-charts";

const chartOptions = (width: number) => ({
  layout: {
    background: { type: ColorType.Solid, color: "#001529" },
    textColor: "white",
  },
  grid: {
    vertLines: {
      color: "#303C43",
    },
    horzLines: {
      color: "#303C43",
    },
  },
  localization: {
    dateFormat: "yyyy-MM-dd",
    locale: "en-US",
    priceFormatter: (p: number) => `$ ${p.toFixed(2)}`,
    timeFormatter: (t: string) => dayjs(t).format("YYYY-MM-DD"),
  },
  timeScale: {
    fixLeftEdge: true,
    rightBarStaysOnScroll: true,
    borderVisible: true,
    visible: true,
    minBarSpacing: 0,
    timeVisible: false,
    secondsVisible: false,
    tickMarkFormatter: (t: string) => dayjs(t).format("MMM-DD"),
  },
  width: width,
  height: 400,
});
export default chartOptions;
