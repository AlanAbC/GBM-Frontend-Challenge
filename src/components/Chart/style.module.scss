.chartContainer {
  min-width: 700px;
}
.checkbox {
  margin-top: 1.2em;

  label {
    font-size: 1.2em;
    font-weight: 500;
    color: black;
  }
}
.datePicker {
  margin-bottom: 1.2em;
  font-size: 1em;
  color: black;
  font-weight: bold;
  .rangePicker {
    margin-left: 1em;
  }
}
.displayOption {
  margin-top: 1.2em;
  font-size: 1em;
  color: black;
  font-weight: bold;
  .select {
    margin-left: 1em;
    min-width: 200px;
  }
}

@media screen and (max-width: 767px) {
  .chartContainer {
    min-width: 100%;
  }
  .displayOption {
    width: 100%;
    .select {
      margin-top: 1em;
      margin-left: 0em;
      min-width: 100%;
    }
  }
}
