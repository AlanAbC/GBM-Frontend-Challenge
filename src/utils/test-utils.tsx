import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

const queryClient = new QueryClient();

const reactQueryWrapper = ({ children }: { children: any }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

const mockedFavoritesData = {
  MSFT: {
    key: "MSFT",
    acronym: "NASDAQ",
    date: "2023-11-17T00:00:00+0000",
    high_price: 374.37,
    close_price: 369.85,
    symbol: "MSFT",
    low_price: 367,
    country: "USA",
    volume: 40157000,
    exchange_name: "NASDAQ Stock Exchange",
    name: "Microsoft Corporation",
    open_price: 373.61,
  },
  VOD: {
    close_price: 9.37,
    low_price: 9.32,
    acronym: "NASDAQ",
    name: "Vodafone Group plc",
    exchange_name: "NASDAQ Stock Exchange",
    open_price: 9.35,
    volume: 6237100,
    key: "VOD",
    date: "2023-11-17T00:00:00+0000",
    country: "USA",
    high_price: 9.42,
    symbol: "VOD",
  },
  GOOG: {
    volume: 25590191,
    close_price: 136.94,
    key: "GOOG",
    name: "Alphabet Inc - Class C",
    acronym: "NASDAQ",
    low_price: 135.48,
    open_price: 137.82,
    date: "2023-11-17T00:00:00+0000",
    symbol: "GOOG",
    high_price: 138,
    country: "USA",
    exchange_name: "NASDAQ Stock Exwfchange",
  },
};

export { reactQueryWrapper, mockedFavoritesData };
