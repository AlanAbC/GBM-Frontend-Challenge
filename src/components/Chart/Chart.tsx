"use client";
import { DatePicker, Select, Tag } from "antd";
import {
  createChart,
  ColorType,
  IChartApi,
  ISeriesApi,
} from "lightweight-charts";
import { Dispatch, SetStateAction, useEffect, useRef, useState } from "react";
import styles from "./style.module.scss";
import dayjs from "dayjs";
import { TickerEODDataInterface } from "@/hooks/useGetTickerEODData";
import { colors } from "@/utils/style-utils";
import chartOptions from "./chartOptions";

const displayOptions = [
  { label: "Low Price", value: "low_price" },
  { label: "High Price", value: "high_price" },
  { label: "Open Price", value: "open_price" },
  { label: "Close Price", value: "close_price" },
];

const { RangePicker } = DatePicker;

export interface ChartInterface {
  chartData?: TickerEODDataInterface;
  setChartDateRange: Dispatch<SetStateAction<string[]>>;
}

const Chart = ({ chartData, setChartDateRange }: ChartInterface) => {
  const chartContainerRef = useRef<HTMLDivElement>(null);
  const [chart, setChart] = useState<IChartApi | null>(null);
  const [openLineSeries, setOpenLineSeries] =
    useState<ISeriesApi<"Line"> | null>(null);
  const [closeLineSeries, setCloseLineSeries] =
    useState<ISeriesApi<"Line"> | null>(null);
  const [highLineSeries, setHighLineSeries] =
    useState<ISeriesApi<"Line"> | null>(null);
  const [lowLineSeries, setLowLineSeries] = useState<ISeriesApi<"Line"> | null>(
    null
  );
  const [datePickerOpen, setDatePickerOpen] = useState(false);
  const [chartLoaded, setChartLoaded] = useState(false);

  useEffect(() => {
    // Create the chart when the component mounts
    if (chartContainerRef.current && !chart && !chartLoaded) {
      setChart(
        createChart(
          chartContainerRef?.current,
          chartOptions(chartContainerRef.current.clientWidth)
        )
      );
      setChartLoaded(true);
    }
  }, [chartContainerRef, chart, chartLoaded]);

  useEffect(() => {
    // Update the chart series when new data is available
    if (chartData && chartLoaded && chart) {
      if (openLineSeries) {
        openLineSeries.setData(chartData.open);
      } else {
        const tempopenLineSeries = chart.addLineSeries({
          color: colors["open_price"],
          priceFormat: {
            type: "price",
          },
        });
        tempopenLineSeries.setData(chartData.open);
        setOpenLineSeries(tempopenLineSeries);
      }

      if (closeLineSeries) {
        closeLineSeries.setData(chartData.close);
      } else {
        const tempcloseLineSeries = chart.addLineSeries({
          color: colors["close_price"],
          priceFormat: {
            type: "price",
          },
        });
        tempcloseLineSeries.setData(chartData.close);
        setCloseLineSeries(tempcloseLineSeries);
      }

      if (highLineSeries) {
        highLineSeries.setData(chartData.high);
      } else {
        const temphighLineSeries = chart.addLineSeries({
          color: colors["high_price"],
          priceFormat: {
            type: "price",
          },
        });
        temphighLineSeries.setData(chartData.high);
        setHighLineSeries(temphighLineSeries);
      }

      if (lowLineSeries) {
        lowLineSeries.setData(chartData.low);
      } else {
        const templowLineSeries = chart.addLineSeries({
          color: colors["low_price"],
          priceFormat: {
            type: "price",
          },
        });
        templowLineSeries.setData(chartData.low);
        setLowLineSeries(templowLineSeries);
      }
      chart.timeScale().fitContent();
    }
  }, [
    chartData,
    chartLoaded,
    chart,
    highLineSeries,
    lowLineSeries,
    openLineSeries,
    closeLineSeries,
  ]);

  const hideOrShowLineData = (value: string, hide: boolean) => {
    switch (value) {
      case "low_price":
        if (hide) {
          lowLineSeries?.setData([]);
        } else {
          chartData && lowLineSeries?.setData(chartData?.low);
        }
        break;
      case "high_price":
        if (hide) {
          highLineSeries?.setData([]);
        } else {
          chartData && highLineSeries?.setData(chartData?.high);
        }
        break;
      case "open_price":
        if (hide) {
          openLineSeries?.setData([]);
        } else {
          chartData && openLineSeries?.setData(chartData?.open);
        }
        break;
      case "close_price":
        if (hide) {
          closeLineSeries?.setData([]);
        } else {
          chartData && closeLineSeries?.setData(chartData?.close);
        }
        break;
      default:
        break;
    }
  };

  const handleSelectValue = (newValue: string) => {
    hideOrShowLineData(newValue, false);
  };

  const handleUnselectValue = (removedValue: string) => {
    hideOrShowLineData(removedValue, true);
  };

  return (
    <div className={styles.chartContainer}>
      <div className={styles.datePicker}>
        {` Date range: `}
        <RangePicker
          className={styles.rangePicker}
          defaultValue={[dayjs(), dayjs().subtract(30, "day")]}
          disabledDate={(current) =>
            current.valueOf() >= dayjs().subtract(1, "day").valueOf()
          }
          open={datePickerOpen}
          onOpenChange={setDatePickerOpen}
          onChange={(values) => {
            if (values) {
              const dateRange = [
                dayjs(values[0]).format("YYYY-MM-DD"),
                dayjs(values[1]).format("YYYY-MM-DD"),
              ];
              setChartDateRange(dateRange);
            }
          }}
        />
      </div>
      <div ref={chartContainerRef} />
      <div className={styles.displayOption}>
        {` Display Options: `}
        <Select
          className={styles.select}
          mode="tags"
          size="middle"
          placeholder="Select options to display"
          tagRender={({ value, label }) => (
            //@ts-ignore
            <Tag color={colors[value]}>{label}</Tag>
          )}
          onSelect={handleSelectValue}
          onDeselect={handleUnselectValue}
          options={displayOptions}
          defaultValue={[
            "low_price",
            "high_price",
            "open_price",
            "close_price",
          ]}
        />
      </div>
    </div>
  );
};

export default Chart;
