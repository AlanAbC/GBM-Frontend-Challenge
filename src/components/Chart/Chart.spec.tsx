import React from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  prettyDOM,
} from "@testing-library/react";
import Chart from "./Chart";
import { act } from "react-dom/test-utils";
import "@testing-library/jest-dom";

// Mock the lightweight-charts library
jest.mock("lightweight-charts", () => ({
  createChart: jest.fn(),
  ColorType: {},
  IChartApi: {
    addLineSeries: jest.fn(),
  },
  ISeriesApi: {},
}));

const mockSetDateRange = jest.fn();

const mockChartData = {
  open: [
    {
      time: "2023-10-17",
      value: 330.11,
    },
    {
      time: "2023-10-18",
      value: 331.32,
    },
    {
      time: "2023-10-19",
      value: 326.67,
    },
  ],
  close: [
    {
      time: "2023-10-17",
      value: 330.11,
    },
    {
      time: "2023-10-18",
      value: 331.32,
    },
    {
      time: "2023-10-19",
      value: 326.67,
    },
  ],
  high: [
    {
      time: "2023-10-17",
      value: 330.11,
    },
    {
      time: "2023-10-18",
      value: 331.32,
    },
    {
      time: "2023-10-19",
      value: 326.67,
    },
  ],
  low: [
    {
      time: "2023-10-17",
      value: 330.11,
    },
    {
      time: "2023-10-18",
      value: 331.32,
    },
    {
      time: "2023-10-19",
      value: 326.67,
    },
  ],
};

describe("Chart Component", () => {
  it("renders without crashing", () => {
    const { container } = render(
      <Chart setChartDateRange={mockSetDateRange} />
    );
    expect(container).toMatchSnapshot();
  });
});
