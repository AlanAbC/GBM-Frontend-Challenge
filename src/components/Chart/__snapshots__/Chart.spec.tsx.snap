// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Component renders without crashing 1`] = `
<div>
  <div
    class="chartContainer"
  >
    <div
      class="datePicker"
    >
       Date range: 
      <div
        class="ant-picker ant-picker-range css-dev-only-do-not-override-1adbn6x rangePicker"
      >
        <div
          class="ant-picker-input ant-picker-input-active"
        >
          <input
            autocomplete="off"
            placeholder="Start date"
            readonly=""
            size="12"
            value="2023-10-19"
          />
        </div>
        <div
          class="ant-picker-range-separator"
        >
          <span
            aria-label="to"
            class="ant-picker-separator"
          >
            <span
              aria-label="swap-right"
              class="anticon anticon-swap-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="swap-right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M873.1 596.2l-164-208A32 32 0 00684 376h-64.8c-6.7 0-10.4 7.7-6.3 13l144.3 183H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h695.9c26.8 0 41.7-30.8 25.2-51.8z"
                />
              </svg>
            </span>
          </span>
        </div>
        <div
          class="ant-picker-input"
        >
          <input
            autocomplete="off"
            placeholder="End date"
            readonly=""
            size="12"
            value="2023-11-18"
          />
        </div>
        <div
          class="ant-picker-active-bar"
          style="left: 0px; width: 0px; position: absolute;"
        />
        <span
          class="ant-picker-suffix"
        >
          <span
            aria-label="calendar"
            class="anticon anticon-calendar"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="calendar"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
              />
            </svg>
          </span>
        </span>
        <span
          class="ant-picker-clear"
          role="button"
        >
          <span
            aria-label="close-circle"
            class="anticon anticon-close-circle"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="close-circle"
              fill="currentColor"
              fill-rule="evenodd"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div />
    <div
      class="displayOption"
    >
       Display Options: 
      <div
        class="ant-select select css-dev-only-do-not-override-1adbn6x ant-select-multiple ant-select-show-arrow ant-select-show-search"
      >
        <div
          class="ant-select-selector"
        >
          <div
            class="ant-select-selection-overflow"
          >
            <div
              class="ant-select-selection-overflow-item"
              style="opacity: 1;"
            >
              <span>
                <span
                  class="ant-tag ant-tag-has-color css-dev-only-do-not-override-1adbn6x"
                  style="background-color: rgb(209, 92, 84);"
                >
                  Low Price
                </span>
              </span>
            </div>
            <div
              class="ant-select-selection-overflow-item"
              style="opacity: 1;"
            >
              <span>
                <span
                  class="ant-tag ant-tag-has-color css-dev-only-do-not-override-1adbn6x"
                  style="background-color: rgb(108, 179, 63);"
                >
                  High Price
                </span>
              </span>
            </div>
            <div
              class="ant-select-selection-overflow-item"
              style="opacity: 1;"
            >
              <span>
                <span
                  class="ant-tag ant-tag-has-color css-dev-only-do-not-override-1adbn6x"
                  style="background-color: rgb(41, 98, 255);"
                >
                  Open Price
                </span>
              </span>
            </div>
            <div
              class="ant-select-selection-overflow-item"
              style="opacity: 1;"
            >
              <span>
                <span
                  class="ant-tag ant-tag-has-color css-dev-only-do-not-override-1adbn6x"
                  style="background-color: rgb(163, 155, 7);"
                >
                  Close Price
                </span>
              </span>
            </div>
            <div
              class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
              style="opacity: 1;"
            >
              <div
                class="ant-select-selection-search"
                style="width: 0px;"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  role="combobox"
                  type="search"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="ant-select-selection-search-mirror"
                >
                  Â 
                </span>
              </div>
            </div>
          </div>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
